src/cluster.c

```diff
@@ -308,8 +432,27 @@ int clusterLoadConfig(char *filename) {
                 master = createClusterNode(argv[3],0);
                 clusterAddNode(master);
             }
+            /* shard_id can be absent if we are loading a nodes.conf generated
+             * by an older version of Redis; we should follow the primary's
+             * shard_id in this case */
+            if (auxUpdaters[shard_id_idx].present == 0) {
+                memcpy(n->shard_id, master->shard_id, CLUSTER_NAMELEN);
+                clusterAddNodeToShard(master->shard_id, n);
```

Comment:
This doesn't seem quite right. The master's shard id might have been automatically generated, up on line 433, and will have a random shard_id. The replica will be assigned the random shard_id, but the master may have its ID actually set later.

Generally it seems like we should resolve these shard_ids after we're done loading all of the nodes and we can also do the validation then.

Reply:
>The master's shard id might have been automatically generated, up on line 433, and will have a random shard_id. 

Correct.

>The replica will be assigned the random shard_id, but the master may have its ID actually set later.

No. The replica will be inheriting the primary's randomly generated shard_id on line 440 even when the loading logic hasn't encountered the definition for the primary node on line 291 and 456. Later on, the loading logic will first hit line 291 where it will find out a `clusterNode` has been created already (by line 433). It then sees "-" and takes the branch on line 451 to add itself to the shard created by one of its replicas. The primary will not reset its shard_id.